app/Http/Controllers/CourseController.php

```diff
@@ -0,0 +1,203 @@
+<?php
+
+namespace App\Http\Controllers;
+
+use Carbon\Carbon;
+use App\Models\Course;
+use Illuminate\Http\Request;
+use Illuminate\Support\Facades\DB;
+use Illuminate\Support\Facades\Cache;
+use Symfony\Component\HttpFoundation\Response;
+
+
+class CourseController extends Controller
+{   
+    private $courseData;
+
+    public function __construct()
+    {   //database取出來的就是一堆物件，裡面有一個叫做name的屬性，可以直接拿出來用
+        //DB::table('course')->get('name') //原本使用這個操作
+        //但是現在改用模型操作，模型有指定table
+        /**
+         * 減少資料庫查詢的次數，所以在建構子就把需要用到的屬性都放進陣列中
+         * 
+         * 
+         */
+
+        // foreach(Course::all() as $obj){
+        //     $this->courseData[$obj->course_id] = [
+        //         'id' => $obj->id,
+        //         'name' => $obj->name,
+        //         'description' => $obj->description,
+        //         'outline' => $obj->outline, 
+        //         'created_at' => $obj->created_at,
+        //         'updated_at' => $obj->updated_at,
+        //     ];
+        // }
+        //$arr[8]["students"][0]["course_id"]
+        foreach(Course::with('students')->get() as $obj){
+            if(count($obj->students) == 0){
+                $this->courseData[$obj->course_id] = [
+                    'id' => $obj->id,
+                    'name' => $obj->name,
+                    'description' => $obj->description,
+                    'outline' => $obj->outline, 
+                    'created_at' => $obj->created_at,
+                    'updated_at' => $obj->updated_at,
+                    //'students' => $obj->students[0]["first_name"]." ".$obj->students[0]["last_name"],
+                    'students' => "Seems that nobody wanna take this QQ",
+                ];
+            } else {
+                $this->courseData[$obj->course_id] = [
+                    'id' => $obj->id,
+                    'name' => $obj->name,
+                    'description' => $obj->description,
+                    'outline' => $obj->outline, 
+                    'created_at' => $obj->created_at,
+                    'updated_at' => $obj->updated_at,
+                    'students' => $obj->students,
+                ];
+                
+            }
```

Comment:
可以這樣寫更精簡
```suggestion
            $this->courseData[$obj->course_id] = [
                'id' => $obj->id,
                'name' => $obj->name,
                'description' => $obj->description,
                'outline' => $obj->outline, 
                'created_at' => $obj->created_at,
                'updated_at' => $obj->updated_at,
                'students' => $obj->students ??  'Seems that nobody wanna take this QQ',
            ];

```
btw 因為取回的格式已經是 collection ，也可以使用  collection  相關的 method 
另外這邊確定有練習到 eager loading 

但有另一個情境每次 request 只有取一個課程，所以在一開始只取需要的資料會更好
這邊看起來可以 搜尋 course_id = $course_id 達成
另外 course_id 比較像 short_name

https://www.php.net/manual/en/migration70.new-features.php#migration70.new-features.null-coalesce-op